File: src/variables.rs
use crate::context::*;
use crate::debug_adapter_comms;
use crate::types::{Scope,Variable};
use crate::kakoune;

use json::object;

//Handles the "scopes" response.
pub fn handle_scopes_response(msg: json::JsonValue, ctx: &mut Context) {
    //Update the "scopes" array in the context
    ctx.scopes.clear();
    ctx.variables.clear();
    let scopes_members = msg["body"]["scopes"].members();
    for val in scopes_members {
        let value = val.clone();
        //Enter this scope in the scopes array
        let scope = Scope {
            variable_reference: val["variablesReference"].to_string().parse::<u64>().unwrap(),
            line_no: 0,
            contents: value,
        };
        ctx.scopes.push(scope);
        //Request variables for this scope
        let var_ref = val["variablesReference"].to_string().parse::<u64>().unwrap();
        let var_args = object!{
            "variablesReference": var_ref
        };
        ctx.var_reqs += 1;
        debug_adapter_comms::do_request("variables".to_string(), var_args, ctx);
    }
}

//Handles the "variables" response.
pub fn handle_variables_response(msg: json::JsonValue, ctx: &mut Context) {
    ctx.var_reqs -= 1;
    //Find the variables request that spawned this response
    let cur_requests = &ctx.cur_requests.clone();
    let val_req = cur_requests.into_iter().find(|x| &x["seq"] == &msg["request_seq"]).unwrap();

    //Loop over every variable in this response
    let variables = &msg["body"]["variables"];
    let variables_members = variables.members();
    for val in variables_members {
        let val_cln = val.clone();
        //Construct an Expandable instance containing this variable's properties
        let variable = Variable {
            variable_reference: val["variablesReference"].to_string().parse::<u64>().unwrap(),
            par_variable_reference: val_req["arguments"]["variablesReference"].to_string().parse::<u64>().unwrap(),
            line_no: 0,
            contents: val_cln,
        };
        ctx.variables.push(variable);
    }
    // If we've serviced all pending variable requests, render the scopes and variables in the variables buffer
    if ctx.var_reqs == 0 {
        serialize_variables(ctx);
    }
}

//Constructs the command that renders all the scopes and variables in the variables buffer.
pub fn serialize_variables(ctx: &mut Context) {
    let mut cmd = "dap-show-variables 'Variables:\n\n".to_string();
    let mut cmd_val = "".to_string();
    for scope in &ctx.scopes {
        let scope_name = &scope.contents["name"];
        cmd_val.push_str(&"Scope: ".to_string());
        cmd_val.push_str(&scope_name.to_string());
        cmd_val.push_str("\n");
        // First confirm that this scope has child variables
        let mut has_child = false;
        for var in &ctx.variables {
            if var.par_variable_reference == scope.variable_reference {
                has_child = true;
                break;
            }
        }
        if has_child {
            let val = serialize_variable(ctx, scope.variable_reference, 4);
            cmd_val.push_str(&val);
        }
    }
    cmd.push_str(&kakoune::editor_escape(&cmd_val));
    cmd.push_str("'");
    kakoune::kak_command(cmd, ctx);
}

//Constructs the command that renders all the child variables of the given variable reference in the variables buffer.
pub fn serialize_variable(ctx: &Context, par_ref: u64, indent: u64) -> String {
    let mut val = "".to_string();
    for var in &ctx.variables {
        if var.par_variable_reference == par_ref {
            for _i in 0..indent {
                val.push_str(" ");
            }
            //If this variable is expandable
            if var.variable_reference > 0 {
                let mut icon = "+";
                //Determine if this variable has any child variables currently
                for varr in &ctx.variables {
                    if varr.par_variable_reference == var.variable_reference {
                        icon = "-";
                        break;
                    }
                }
                val.push_str(&format!("{} ", icon));
            }
            val.push_str("<");
            val.push_str(&var.variable_reference.to_string());
            val.push_str("> ");
            val.push_str(&var.contents["name"].to_string());
            val.push_str(" (");
            val.push_str(&var.contents["type"].to_string());
            val.push_str("): ");
            val.push_str(&var.contents["value"].to_string());
            val.push_str("\n");
            if var.variable_reference > 0 {
                val.push_str(&serialize_variable(ctx, var.variable_reference, indent + 4));
            }
        }
    }
    val
}

//Handles the "expand" command from the editor.
pub fn expand_variable(cmd: &String, ctx: &mut Context) {
    let mut var = cmd.to_string();
    var = var.trim().to_string();
    //If the string starts with a '<', this is an expandable variable
    let first_char = var.chars().next().unwrap();
    if first_char == '<' {
        //Extract the variable reference
        var = var[1..].to_string();
        let var_ref = var.parse::<u64>().unwrap();
        //If the variables list contains any child variables of this variable reference, then it's expanded
        let mut is_expanded = false;
        for varr in &ctx.variables {
            if varr.par_variable_reference == var_ref {
                is_expanded = true;
                break;
            }
        }
        //If this variable isn't expanded, then expand it
        if !is_expanded {
            let var_args = object!{
                "variablesReference": var_ref
            };
            ctx.var_reqs += 1;
            debug_adapter_comms::do_request("variables".to_string(), var_args, ctx);
        }
        //Otherwise, collapse it
        else {
            ctx.variables.retain(|x| x.par_variable_reference != var_ref);
            serialize_variables(ctx);
        }
    }
}

File: src/variables.py
import debug_session

from functools import partial

scopes = []
pending_var_requests = 0
line_to_var = {}
cur_line = 0


class Expandable:
    def __init__(self):
        self.variable_reference = 0
        self.children = []
        self.contents = {}


def handle_scopes_response(msg):
    global scopes, pending_var_requests
    # TODO: Make scopes and variables persistent
    scopes.clear()

    scopes_members = msg["body"]["scopes"]
    for val in scopes_members:
        scope = Expandable()
        scope.variable_reference = int(val["variablesReference"])
        scope.contents = val
        scopes.append(scope)
        # Request variables for this scope
        var_ref = int(val["variablesReference"])
        var_args = {"variablesReference": var_ref}
        pending_var_requests += 1
        debug_session.debug_adapter.write_request(
            "variables", var_args, partial(handle_variables_response, parent=scope)
        )


def handle_variables_response(msg, parent):
    global variables, pending_var_requests

    pending_var_requests -= 1

    # Loop over every variable in this response
    var_members = msg["body"]["variables"]
    for val in var_members:
        variable = Expandable()
        variable.variable_reference = int(val["variablesReference"])
        variable.contents = val
        parent.children.append(variable)

    # If we've serviced all pending variable requests, render the scopes and variables
    if pending_var_requests == 0:
        serialize_scopes()


def serialize_scopes():
    global cur_line, line_to_var
    line_to_var.clear()

    cmd = "dap-show-variables '"
    cmd_val = ""
    cur_line = 1
    for scope in scopes:
        scope_name = scope.contents["name"]
        cmd_val += f"Scope: {scope_name}"
        cmd_val += "\n"
        cur_line += 1
        # If this scope has child variables, render them
        if len(scope.children) != 0:
            cmd_val += serialize_variables(scope.children, 2)

    cmd += debug_session.kak_connection.escape_str(cmd_val)
    cmd += "'"
    debug_session.kak_connection.send_cmd(cmd)


def serialize_variables(variables, indent):
    global cur_line, line_to_var
    val = ""
    icon = " "  # + if collapsed, - if expanded, ' ' otherwise
    for var in variables:
        line_to_var[cur_line] = var
        # Indent
        for i in range(0, indent):
            val += " "
        # Determine proper icon
        if var.variable_reference > 0:
            icon = "+"
            if len(var.children) != 0:
                icon = "-"
        # Render variable
        var_name = var.contents["name"]
        var_type = var.contents["type"]
        var_value = var.contents["value"]
        val += f"{icon} "
        val += f"{var_name} ({var_type}): {var_value}\n"
        cur_line += 1
        # If variable is expanded, render children
        if len(var.children) != 0:
            val += serialize_variables(var.children, indent + 2)
    return val


def expand_variable(line):
    global pending_var_requests
    # If this is a line with a variable, then expand it
    if line in line_to_var.keys():
        parent = line_to_var[line]
        # If the variable has children, it's expanded, so collapse it
        if len(parent.children) != 0:
            parent.children.clear()
            serialize_scopes()
        elif parent.variable_reference > 0:
            # This isn't expanded, but is expandable. Expand it.
            var_args = {"variablesReference": parent.variable_reference}
            pending_var_requests += 1
            debug_session.debug_adapter.write_request(
                "variables", var_args, partial(handle_variables_response, parent=parent)
            )
